### DISCUSSION TITLE
[java] False-positive for Objects method when static import present?

### DISCUSSION THREAD

**Discussion by linusjf** [pmd/pmd#3082] [General]
When I have a static method import with the same name as a compilation unit's method and the compilation unit's method(s) do not match the imported method's signature, then the java compiler fails complaining:

"reason: actual and formal argument lists differ in length"

This is mitigated by invoking the method call using the fully qualified name. 

PMD, however, is able to find the static method and does not crib---it is apparently smarter than the java compiler. However, it flags the FQN method call as a violation of the rule UnnecessaryFullyQualifiedName even though the java compiler mandates it.

On digging further on Google and StackOverflow, I discovered the following post:

https://stackoverflow.com/questions/7890853/why-cant-i-static-import-an-equals-method-in-java/28749909#28749909

Now, the question is:

Is it worth PMD modifying the rule to suppress Object class method overloads in static methods i.e., if a FQN static method overloads an Object class method it can be suppressed? Or should there be a customizable field that lists FQMNs to be ignored by this rule? 

I can live with  suppressing the violation.

```java

package ds;
 import static java.util.Objects.requireNonNull;
import static java.util.Objects.isNull;
import static java.util.Objects.nonNull;
import static java.util.Objects.toString; 

import java.util.Iterator; 

public class BinaryTree<E extends Comparable<E>> implements Tree<E> { 

private TreeNode<E> root;

 public BinaryTree() { 
root = null; 

} 

 @SuppressWarnings("nullness") 
@Override
 public void add(E obj) { 
requireNonNull(obj); 
if (isNull(root))
 root = new TreeNode<>(obj, null, null); 
else 
root.insert(obj);
 }
 @SuppressWarnings("PMD.SimplifiedTernary") @Override 
public boolean contains(E obj) { 
requireNonNull(obj);
 return nonNull(root) ? nonNull(root.find(obj)) : false;
 } 

@Override 
public void remove(E obj) { 
if (nonNull(root)) 
root = root.remove(obj, root);
 }

 @Override
 public Iterator<E> iterator(TraversalOrder order) {
 return null; 
} 

private static class TreeNode<T extends Comparable<T>> {
 T val; 
TreeNode<T> left; 
TreeNode<T> right; 

 @SuppressWarnings("nullness") 
TreeNode(T val, TreeNode<T> left, TreeNode<T> right) {
 this.val = val; 
this.left = left;
 this.right = right;
 } 

 @SuppressWarnings("nullness") 
public void insert(T obj) {
 requireNonNull(obj); 
if (val.compareTo(obj) < 0) { 
if (isNull(right)) right = new TreeNode<>(obj, null, null); 
else right.insert(obj); } 
else { 
if (isNull(left))
 left = new TreeNode<>(obj, null, null);
 else left.insert(obj);
 } 
} 
 @SuppressWarnings("nullness")
 public TreeNode<T> find(T obj) { 
int temp = val.compareTo(obj); 
if (temp == 0) return this; 
if (temp < 0) return isNull(right) ? null : right.find(obj); 
return isNull(left) ? null : left.find(obj);

 } 
 @SuppressWarnings({"PMD.LawOfDemeter", "nullness"})
 private TreeNode<T> remove(T obj, TreeNode<T> node) { 
requireNonNull(obj);
 TreeNode<T> t = node; 
if (isNull(t)) return t; 
if (obj.compareTo(t.val) < 0) t.left = remove(obj, t.left); 
else if (obj.compareTo(t.val) > 0) t.right = remove(obj, t.right); 
else if (isNull(t.left) || isNull(t.right)) t = isNull(t.left) ? t.right : t.left; 
else { t.val = findMin(t.right).val; t.right = remove(t.val, t.right);
 } return t;
 } 

 @SuppressWarnings("nullness")
 private TreeNode<T> findMin(TreeNode<T> node) { 
TreeNode<T> t = node; 
if (isNull(t)) return null; 
while (nonNull(t.left)) 
t = t.left; 
return t; 
}
 
 @Override 
public String toString() { 
return toString(val);
 } 
} 
}
```

**Comment by oowekyala** [comment]
This is incorrect behavior by the rule. The rule should not be resolving the method when javac does not. I think it doesn't understand that inherited methods shadow imported methods. This is a more general problem than just the methods of Object.

Adding a property with exceptions is definitely inappropriate. The shadowing behavior should be fixed instead.

**Reply by linusjf** [reply]
Written up as [false-positive](https://github.com/pmd/pmd/issues/3083)
